pyfr/plugins/acoustics.py

```diff
@@ -0,0 +1,196 @@
+
+    def _distances(self, etype, eles, fslice):
+        spts = eles.ploc_at_np('fpts').transpose(0, 2, 1)[fslice]
+        s = spts.reshape((-1, self.ndims))
+        o = self.obsv_pts
+
+        r_o = o[None, :] - s[:, None]
+        d = np.linalg.norm(r_o, axis=-1)
+        x = r_o / d[:, :, None]
+
+        m_r = self.qinf['M']*x
+        m_r_dot = np.sum(m_r, axis=-1)
+
+        inv_gamma = np.sqrt(1 - self.Minf**2)
+
+        gamma = 1 / inv_gamma
+        s_fact = np.sqrt(inv_gamma*inv_gamma + m_r_dot*m_r_dot)
+        r_star_vec = r_o*s_fact[..., None]
+        r_vec = gamma*gamma*(r_star_vec - r_o*m_r_dot[..., None])
+
+        r_star = np.linalg.norm(r_star_vec, axis=-1)
+        r = np.linalg.norm(r_vec, axis=-1)
+
+        r_star_vec_n = r_star_vec / r_star[..., None]
+        r_vec_n = r_vec / r[..., None]
+
+        return r_vec_n, 1 / r_star, r_star_vec_n
+
+    def fwh_solve(self, intg):
+        o_vals = np.zeros(self.nobvs)
+        ci = 1 / self.qinf['c']
+
+        soln, soln_t = {}, {}
+        # Querry solutions all at once to prevent MPI deadlock
+        for (etype, fidx), p in self.proj.items():
+            eidx, m0, n, a, (R, r_star_inv, R_star) = p
+            s = intg.soln[self.emap[etype]][..., eidx]
+            soln[etype, fidx] = m0 @ s.transpose(1, 0, 2)
+
+            s_t = intg.dt_soln[self.emap[etype]][..., eidx]
+            soln_t[etype, fidx] = m0 @ s_t.transpose(1, 0, 2)
+
+        # Accumulate FWH contribution from each surface part
+        for (etype, fidx), p in self.proj.items():
+            eidx, m0, n, a, (R, r_star_inv, R_star) = p
+
+            s, s_t = soln[etype, fidx], soln_t[etype, fidx]
+
+            pris = self.elementscls.con_to_pri(s, self.cfg)
+            pris = np.reshape(pris, (self.nvars, -1))
+
+            p = pris[self._pidx] - self.qinf['p']
+            u = pris[self._vidx] - self.uinf[:, None]
+            r_inf = self.qinf['rho']
+            r_tot = r_inf + p*ci**2
+
+            mom = r_tot*(u + self.uinf[:, None])
+            drift = -r_inf*self.uinf[:, None]
+
+            q = np.einsum('ij,ij->j', n, mom + drift)
+
+            mom_n = np.einsum('ij,ij->j', n, mom)
+            F = mom_n[None, ...]*u + (p - self.qinf['p'])*n
+
+            # Time derivatives of q and F
+            pris_t = self.elementscls.diff_con_to_pri(s, s_t, self.cfg)
+            pris_t = np.reshape(pris_t, (self.nvars, -1))
+
+            u_t = pris_t[self._vidx]
+            p_t = pris_t[self._pidx]
+
+            r_tot_t = p_t / (self.qinf['c']**2)
+            mom_t = (r_tot_t*(u + self.uinf[:, None]) + r_tot*u_t)
+            mom_t_n = np.einsum('ij,ij->j', n, mom_t)
+
+            q_t = mom_t_n
+            F_t = mom_t_n[None, ...]*u + mom_n[None, ...]*u_t + p_t*n
+
+            # Thickness noise contribution
+            I = (1 - self.Minf)*q_t[:, None]*r_star_inv
+            I -= (R_star @ self.uinf)*q[:, None]*r_star_inv**2
```

Comment:
Would it be legitimate to multiply R_star through by r_star_inv into a temporary?